impl/maven-impl/src/test/java/org/apache/maven/impl/model/DefaultModelProcessorTest.java

```diff
@@ -0,0 +1,184 @@
+    }
+
+    @Test
+    void locateExistingPomParserReturnsPathOutsideProjectShouldThrow() {
+        Path unexpectedPom = Path.of("other/pom.xml");
+        Source mockSource = mock(Source.class);
+        when(mockSource.getPath()).thenReturn(unexpectedPom);
+
+        ModelParser parser = mock(ModelParser.class);
+        when(parser.locate(any())).thenReturn(Optional.of(mockSource));
+
+        DefaultModelProcessor processor = new DefaultModelProcessor(mock(ModelXmlFactory.class), List.of(parser));
+        IllegalArgumentException ex =
+                assertThrows(IllegalArgumentException.class, () -> processor.locateExistingPom(Path.of("project")));
+
+        assertTrue(ex.getMessage().contains("does not belong to the given directory"));
+    }
+
+    @Test
+    void readWithAllParsersAndFactoryFailShouldThrowIOExceptionWithSuppressed() throws Exception {
+        ModelXmlFactory factory = mock(ModelXmlFactory.class);
+        ModelParser parser = mock(ModelParser.class);
+        XmlReaderRequest request = mock(XmlReaderRequest.class);
+        Path path = Path.of("project/pom.xml");
+
+        when(request.getPath()).thenReturn(path);
+        when(request.isStrict()).thenReturn(true);
+        when(parser.locateAndParse(any(), any())).thenThrow(new ModelParserException("Parser failed"));
+
+        IOException factoryException = new IOException("Factory failed");
+        when(factory.read(request)).thenThrow(factoryException);
+
+        DefaultModelProcessor processor = new DefaultModelProcessor(factory, List.of(parser));
+
+        IOException thrown = assertThrows(IOException.class, () -> processor.read(request));
+        assertEquals("Factory failed", thrown.getMessage());
+        Throwable[] suppressed = thrown.getSuppressed();
+        assertEquals(1, suppressed.length);
+        assertInstanceOf(ModelParserException.class, suppressed[0]);
+        assertEquals("Parser failed", suppressed[0].getMessage());
+    }
+
+    @Test
+    void locateExistingPomFallbackWithValidPomShouldReturnPom() throws Exception {
+        Path projectDir = Files.createTempDirectory("testproject");
+        Path pomFile = projectDir.resolve("pom.xml");
+        Files.createFile(pomFile);
+
+        DefaultModelProcessor processor = new DefaultModelProcessor(mock(ModelXmlFactory.class), List.of());
+        Path result = processor.locateExistingPom(projectDir);
+
+        assertEquals(pomFile, result);
+
+        Files.deleteIfExists(pomFile);
+        Files.deleteIfExists(projectDir);
+    }
+
+    @Test
+    void locateExistingPomFallbackWithFileAsPathShouldReturnThatFile() throws Exception {
+        Path tempPom = Files.createTempFile("pom", ".xml");
+
+        DefaultModelProcessor processor = new DefaultModelProcessor(mock(ModelXmlFactory.class), List.of());
+        Path result = processor.locateExistingPom(tempPom);
+
+        assertEquals(tempPom, result);
+
+        Files.deleteIfExists(tempPom);
```

Comment:
cleanup probably belongs in a finally block, or perhaps an @After method

Reply:
yes its DRY